cub/cub/agent/agent_find.cuh

```diff
@@ -0,0 +1,306 @@
+  // Per-thread fields
+  //---------------------------------------------------------------------
+
+  _TempStorage& sresult; ///< Reference to temp_storage
+  InputIteratorT d_in; ///< Input data to find
+  // OutputIteratorT d_out;
+  // OffsetT num_items;
+  // OffsetT* value_temp_storage;
+  // WrappedInputIteratorT d_wrapped_in; ///< Wrapped input data to find
+  ScanOpT scan_op; ///< Binary reduction operator
+
+  //---------------------------------------------------------------------
+  // Utility
+  //---------------------------------------------------------------------
+
+  template <typename T>
+  static _CCCL_DEVICE _CCCL_FORCEINLINE bool
+  IsAlignedAndFullTile(T* d_in, int tile_offset, int tile_size, OffsetT num_items, Int2Type<true> /*CAN_VECTORIZE*/)
+  {
+    /// Create an AgentFindIf and extract these two as type member in the encapsulating struct
+    using InputT  = T;
+    using VectorT = typename CubVector<InputT, VECTOR_LOAD_LENGTH>::Type;
+    ///
+    const bool full_tile  = (tile_offset + tile_size) <= num_items;
+    const bool is_aligned = reinterpret_cast<::cuda::std::uintptr_t>(d_in) % uintptr_t{sizeof(VectorT)} == 0;
+    return full_tile && is_aligned;
+  }
+
+  template <typename Iterator>
+  static _CCCL_DEVICE _CCCL_FORCEINLINE bool IsAlignedAndFullTile(
+    Iterator /*d_in*/,
+    int /*tile_offset*/,
+    int /*tile_size*/,
+    std::size_t /*num_items*/,
+    Int2Type<false> /*CAN_VECTORIZE*/)
+  {
+    return false;
+  }
+
+  //---------------------------------------------------------------------
+  // Constructor
+  //---------------------------------------------------------------------
+
+  /**
+   * @brief Constructor
+   * @param sresult Reference to temp_storage
+   * @param d_in Input data to search
+   * @param scan_op Binary scan operator
+   */
+  _CCCL_DEVICE _CCCL_FORCEINLINE AgentFind(TempStorage& sresult, InputIteratorT d_in, ScanOpT scan_op)
+      : sresult(sresult.Alias())
+      , d_in(d_in)
+      , scan_op(scan_op)
+  {}
+
+  //---------------------------------------------------------------------
+  // Tile consumption
+  //---------------------------------------------------------------------
+
+  template <typename Pred>
+  __device__ void
+  ConsumeTile(int tile_offset, Pred pred, OffsetT* result, OffsetT num_items, Int2Type<true> /*CAN_VECTORIZE*/)
+  {
+    using InputT  = cub::detail::value_t<InputIteratorT>;
+    using VectorT = typename CubVector<InputT, VECTOR_LOAD_LENGTH>::Type;
+
+    __shared__ OffsetT block_result;
+
+    if (threadIdx.x == 0)
+    {
+      block_result = num_items;
+    }
+
+    __syncthreads();
+
+    constexpr int NUMBER_OF_VECTORS = ITEMS_PER_THREAD / VECTOR_LOAD_LENGTH;
+    //// vectorized loads begin
+    InputT* d_in_unqualified = const_cast<InputT*>(d_in) + tile_offset + (threadIdx.x * VECTOR_LOAD_LENGTH);
+
+    cub::CacheModifiedInputIterator<AgentFindPolicy::LOAD_MODIFIER, VectorT> d_vec_in(
+      reinterpret_cast<VectorT*>(d_in_unqualified));
```

Comment:
Suggestion: Just checked `CacheModifiedInputIterator`. It's constructor can take a qualified version of the value type (which it strips internally). You should be able to pass a const-qualified pointer:
```suggestion
    const InputT* d_in_unqualified = d_in + tile_offset + (threadIdx.x * VECTOR_LOAD_LENGTH);

    cub::CacheModifiedInputIterator<AgentFindPolicy::LOAD_MODIFIER, VectorT> d_vec_in(
      reinterpret_cast<const VectorT*>(d_in_unqualified));
```
If this does not work, please tell me and we investigate the reason.